src/H5Tconv.c

```diff
@@ -2032,17 +2042,27 @@ H5T__conv_struct_init(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata)
                 } /* end if */
             }     /* end for */
             if (src2dst[i] >= 0) {
-                hid_t  tid;
                 H5T_t *type;
+                hid_t  tid;
 
-                type = H5T_copy(src->shared->u.compnd.memb[i].type, H5T_COPY_ALL);
-                tid  = H5I_register(H5I_DATATYPE, type, false);
-                assert(tid >= 0);
+                if (NULL == (type = H5T_copy(src->shared->u.compnd.memb[i].type, H5T_COPY_ALL)))
+                    HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCOPY, FAIL,
+                                "can't copy source compound member datatype");
+                priv->src_memb[i] = type;
+
+                if ((tid = H5I_register(H5I_DATATYPE, type, false)) < 0)
```

Comment:
Why do we need IDs for these types still?

Reply:
Still need them for passing to the lower members functions in case a conversion exception function is being used. I'd like to only conditionally allocate the arrays and create the IDs, but there currently isn't anything passed to the conversion function that could tell me at init time whether I need to. The new conversion context is only available at conversion time, not init time. I suppose we could add a field to the `H5T_cdata_t` structure, though since it's a public one I've tried to avoid that. It also wouldn't really be of interest to an application conversion callback either.

Reply:
Couldn't you call H5T_convert here, instead of H5T_convert_ctx, and let it sort things out?

Reply:
Or add a helper routine (maybe in the _init() call) that determined which ones needed an ID and which could avoid making one?

Reply:
The ctx version of convert was actually created for this specific reason since calling H5T_convert would cause an enormous amount of overhead as it would be creating and destroying IDs for every member of every compound element when a conversion exception function is used. That said, maybe I can provide the context object at init time with a field that let's the function know if it needs these IDs and then separate which fields of the context will be valid based on the conversion command type specified in the cdata.

Reply:
Yes, that's the direction I was thinking.